### PR TITLE
Ecdsa/Secp256k1 implementation

### PR DESCRIPTION
**Pull request by justkawal**
## **User description**
Add Ecdsa/Secp256k1 usage functionality:
- Sign/Verify Messages
- Integrate to Keypairs
- Add a new library from ground up for secp256k1 ( Made from group up from Noble-curves reference ).
- Adds ecdsa `Hard key derivation` in substrate-bip39.
- Test wide range of cases on Ecdsa/secp256k1.

Note: (Test Resource json files are big)

___

## **Type**
Enhancement

___

## **Description**
- Added a new `Utilities` class in `secp256k1_ecdsa` package that provides various utility functions for the ECDSA/Secp256k1 implementation.
- Added a new `json_model.dart` file in `secp256k1_ecdsa` package that contains various classes for handling JSON data in tests.
- Updated the `Keyring` class in `polkadart_keyring` package to support ECDSA.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>utilities.dart</strong><dd><code>Addition of Utilities class for ECDSA/Secp256k1 implementation</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/secp256k1_ecdsa/lib/utils/utilities.dart
- Added a new `Utilities` class that provides various utility functions <br>  for the ECDSA/Secp256k1 implementation.
- The class includes functions for generating random BigInt, converting <br>  hexadecimal to bytes, modular inversion, bytes to BigInt, BigInt to <br>  bytes, HMAC-DRBG, and many more.
    </details>
  </td>
  <td><a href="https://github.com/leonardocustodio/polkadart/pull/404/files#diff-87c2fb3370a910caaaaa5366973304ecb234fc7ca50c170759bc994d6aa007f0">+352/-0</a>&nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>keyring.dart</strong><dd><code>Update Keyring class to support ECDSA</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/polkadart_keyring/lib/src/keyring.dart
- Added a new `Keyring` instance for ECDSA.
- Updated the `fromSeed` method to handle ECDSA key pair type.
    </details>
  </td>
  <td><a href="https://github.com/leonardocustodio/polkadart/pull/404/files#diff-8c3f38ea3b853d363e214cbdbffa8bc592811011af958f2d78a38d93c9cf9cdc">+13/-3</a>&nbsp; &nbsp; </td>
</tr>                    
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>json_model.dart</strong><dd><code>Addition of JSON model classes for test data handling</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/secp256k1_ecdsa/test/helpers/json_model.dart
- Added a new `json_model.dart` file that contains various classes for <br>  handling JSON data in tests.
- Classes include `PointModel`, `Valid`, `InValid`, `PointMultiply`, `<br>  ``PointFromScalar`, `<br>  ``PointCompress`, `<br>  ``PointAdd`, `<br>  ``PointAddScalar`, `<br>  ``IsPoint`, `<br>  ``IsPrivate`, `<br>  ``PrivateAdd`, `<br>  ``PrivateSub`, `<br>  ``Add`, `<br>  ``Negate`.
    </details>
  </td>
  <td><a href="https://github.com/leonardocustodio/polkadart/pull/404/files#diff-5b02568a6bc341e245c89e90308db35cec4915aee8b63dbc20e4d5821e844f77">+251/-0</a>&nbsp; </td>
</tr>                    
</table></td></tr></tr></tbody></table><hr>

<details> <summary><strong>✨ Usage guide:</strong></summary><hr> 

**Overview:**
The `describe` tool scans the PR code changes, and generates a description for the PR - title, type, summary, walkthrough and labels. The tool can be triggered [automatically](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#github-app-automatic-tools) every time a new PR is opened, or can be invoked manually by commenting on a PR.

When commenting, to edit [configurations](https://github.com/Codium-ai/pr-agent/blob/main/pr_agent/settings/configuration.toml#L46) related to the describe tool (`pr_description` section), use the following template:
```
/describe --pr_description.some_config1=... --pr_description.some_config2=...
```
With a [configuration file](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#working-with-github-app), use the following template:
```
[pr_description]
some_config1=...
some_config2=...
```

<table><tr><td><details> <summary><strong> Enabling\disabling automation </strong></summary><hr>

- When you first install the app, the [default mode](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#github-app-automatic-tools) for the describe tool is:
```
pr_commands = ["/describe --pr_description.add_original_user_description=true" 
                         "--pr_description.keep_original_user_title=true", ...]
```
meaning the `describe` tool will run automatically on every PR, will keep the original title, and will add the original user description above the generated description. 

- Markers are an alternative way to control the generated description, to give maximal control to the user. If you set:
```
pr_commands = ["/describe --pr_description.use_description_markers=true", ...]
```
the tool will replace every marker of the form `pr_agent:marker_name` in the PR description with the relevant content, where `marker_name` is one of the following:
  - `type`: the PR type.
  - `summary`: the PR summary.
  - `walkthrough`: the PR walkthrough.

Note that when markers are enabled, if the original PR description does not contain any markers, the tool will not alter the description at all.
        

</details></td></tr>

<tr><td><details> <summary><strong> Custom labels </strong></summary><hr>

The default labels of the `describe` tool are quite generic: [`Bug fix`, `Tests`, `Enhancement`, `Documentation`, `Other`].

If you specify [custom labels](https://github.com/Codium-ai/pr-agent/blob/main/docs/DESCRIBE.md#handle-custom-labels-from-the-repos-labels-page-gem) in the repo's labels page or via configuration file, you can get tailored labels for your use cases.
Examples for custom labels:
- `Main topic:performance` - pr_agent:The main topic of this PR is performance
- `New endpoint` - pr_agent:A new endpoint was added in this PR
- `SQL query` - pr_agent:A new SQL query was added in this PR
- `Dockerfile changes` - pr_agent:The PR contains changes in the Dockerfile
- ...

The list above is eclectic, and aims to give an idea of different possibilities. Define custom labels that are relevant for your repo and use cases.
Note that Labels are not mutually exclusive, so you can add multiple label categories.
Make sure to provide proper title, and a detailed and well-phrased description for each label, so the tool will know when to suggest it.        

</details></td></tr>

<tr><td><details> <summary><strong> Inline File Walkthrough 💎</strong></summary><hr>

For enhanced user experience, the `describe` tool can add file summaries directly to the "Files changed" tab in the PR page.
This will enable you to quickly understand the changes in each file, while reviewing the code changes (diffs).

To enable inline file summary, set `pr_description.inline_file_summary` in the configuration file, possible values are:
- `'table'`: File changes walkthrough table will be displayed on the top of the "Files changed" tab, in addition to the "Conversation" tab.
- `true`: A collapsable file comment with changes title and a changes summary for each file in the PR.
- `false` (default): File changes walkthrough will be added only to the "Conversation" tab.
<tr><td><details> <summary><strong> Utilizing extra instructions</strong></summary><hr>

The `describe` tool can be configured with extra instructions, to guide the model to a feedback tailored to the needs of your project.

Be specific, clear, and concise in the instructions. With extra instructions, you are the prompter. Notice that the general structure of the description is fixed, and cannot be changed. Extra instructions can change the content or style of each sub-section of the PR description.

Examples for extra instructions:
```
[pr_description] 
extra_instructions="""
- The PR title should be in the format: '<PR type>: <title>'
- The title should be short and concise (up to 10 words)
- ...
"""
```
Use triple quotes to write multi-line instructions. Use bullet points to make the instructions more readable.

</details></td></tr>

<tr><td><details> <summary><strong> More PR-Agent commands</strong></summary><hr> 

> To invoke the PR-Agent, add a comment using one of the following commands:  
> - **/review**: Request a review of your Pull Request.   
> - **/describe**: Update the PR title and description based on the contents of the PR.   
> - **/improve [--extended]**: Suggest code improvements. Extended mode provides a higher quality feedback.   
> - **/ask \<QUESTION\>**: Ask a question about the PR.   
> - **/update_changelog**: Update the changelog based on the PR's contents.   
> - **/add_docs** 💎: Generate docstring for new components introduced in the PR.   
> - **/generate_labels** 💎: Generate labels for the PR based on the PR's contents.   
> - **/analyze** 💎: Automatically analyzes the PR, and presents changes walkthrough for each component.   

>See the [tools guide](https://github.com/Codium-ai/pr-agent/blob/main/docs/TOOLS_GUIDE.md) for more details.
>To list the possible configuration parameters, add a **/config** comment.   
 

</details></td></tr>

</table>

See the [describe usage](https://github.com/Codium-ai/pr-agent/blob/main/docs/DESCRIBE.md) page for a comprehensive guide on using this tool.

</details>

### CODE REVIEW

**Comment by leonardocustodio** [comment]
Can we change the version to `0.1.0` we will bump everything (sr25519/ecdsa) to `1.0.0` when keyring reaches version 1

File: packages/polkadart_keyring/pubspec.yaml [hunk 1]
```diff
@@ -4,15 +4,20 @@ version: 0.3.0
 homepage: https://github.com/leonardocustodio/polkadart/tree/main/packages/polkadart_keyring
 repository: https://github.com/leonardocustodio/polkadart
 
+publish_to: 'none' # Remove this line if you wish to publish to pub.dev
```
Comment (leonardocustodio) [inline packages/polkadart_keyring/pubspec.yaml:7]: You can remove this please.

File: packages/polkadart_keyring/pubspec.yaml [hunk 2]
```diff
@@ -4,15 +4,20 @@ version: 0.3.0
...
 
 # Add regular dependencies here.
 dependencies:
-  cryptography: ^2.5.0
+  pointycastle: ^3.6.2
   flutter_curve25519: ^0.1.2
   ss58: ^1.1.2
-  substrate_bip39: ^0.2.0
+  substrate_bip39:
+    path: ../substrate_bip39
```
Comment (leonardocustodio) [inline packages/polkadart_keyring/pubspec.yaml:18]: Doesn't it work fine when changing in the override? I know for secp256k1 will not work as the package is not published yet.

File: packages/secp256k1_ecdsa/CHANGELOG.md [hunk 3]
```diff
@@ -0,0 +1,3 @@
+## 1.0.0
```
Comment (leonardocustodio) [inline packages/secp256k1_ecdsa/CHANGELOG.md:1]: 0.1.0

File: packages/secp256k1_ecdsa/pubspec.yaml [hunk 4]
```diff
@@ -0,0 +1,16 @@
+name: secp256k1_ecdsa
+description: A starting point for Dart libraries or applications.
+version: 1.0.0
```
Comment (leonardocustodio) [inline packages/secp256k1_ecdsa/pubspec.yaml:3]: 0.1.0

File: packages/secp256k1_ecdsa/pubspec.yaml [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+# repository: https://github.com/my_org/my_repo
```
Comment (leonardocustodio) [inline packages/secp256k1_ecdsa/pubspec.yaml:4]: Can fix the repository here

File: packages/secp256k1_ecdsa/pubspec.yaml [hunk 6]
```diff
@@ -0,0 +1,16 @@
+name: secp256k1_ecdsa
+description: A starting point for Dart libraries or applications.
```
Comment (leonardocustodio) [inline packages/secp256k1_ecdsa/pubspec.yaml:2]: Change the description please

File: packages/secp256k1_ecdsa/pubspec.yaml [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+version: 1.0.0
+# repository: https://github.com/my_org/my_repo
+
+publish_to: 'none' # Remove this line if you wish to publish to pub.dev
```
Comment (leonardocustodio) [inline packages/secp256k1_ecdsa/pubspec.yaml:6]: Remove it

**Review by leonardocustodio — CHANGES_REQUESTED** [review]

File: packages/polkadart_keyring/pubspec.yaml (hunk 1 above)
Comment (justkawal) [reply packages/polkadart_keyring/pubspec.yaml:7]: Removed. ✅

File: packages/secp256k1_ecdsa/pubspec.yaml (hunk 4 above)
Comment (justkawal) [reply packages/secp256k1_ecdsa/pubspec.yaml:3]: Done ✅

File: packages/secp256k1_ecdsa/pubspec.yaml (hunk 5 above)
Comment (justkawal) [reply packages/secp256k1_ecdsa/pubspec.yaml:4]: Done ✅

File: packages/secp256k1_ecdsa/pubspec.yaml (hunk 7 above)
Comment (justkawal) [reply packages/secp256k1_ecdsa/pubspec.yaml:6]: Removed ✅

File: packages/secp256k1_ecdsa/CHANGELOG.md (hunk 3 above)
Comment (justkawal) [reply packages/secp256k1_ecdsa/CHANGELOG.md:1]: Changed ✅

File: packages/polkadart_keyring/pubspec.yaml (hunk 2 above)
Comment (justkawal) [reply packages/polkadart_keyring/pubspec.yaml:18]: Done ✅

**Comment by leonardocustodio** [comment]
Nice job man, I will do some manual testing and I can finish the review by the end of the day.

**Comment by justkawal** [comment]
Let's complete this in next 2 weeks or as early as possible.... Haha

**Review by leonardocustodio — CHANGES_REQUESTED** [review]
There is an issue, I don't think it is generating the correct `public key` here you can find the address generated by polkadot-js using ecdsa:
image
And that's the public key of that address:
image
And here we can see the public key generated by the package:
image

Also important to note that the key generated is 1 byte longer than it should be.
Correct value: `0x0d90aef20b4a64c8831c1420db3a2894d4e0d838d93acaacc3fec2f27b36db4b`
Wrong value: `0x0362cb81f6d8afe001df5b310e10b09f673c3399a63103ec6402ebbcba836619b7`

**Review by leonardocustodio — CHANGES_REQUESTED** [review]
Since we now have `ecdsa` support and thus the ss58 public key and the keyring public key doesn't always match, I think we should add another method to get the `raw address`.

Check the following example from polkadot-js:
```js
const keyring = new Keyring({ ss58Format: 42, type: 'ecdsa' });
const test = keyring.addFromMnemonic('resource mirror lecture smooth midnight muffin position cup pepper fruit vanish also', {});
console.log("Public key:", u8aToHex(test.publicKey));
console.log('SS58 public key', u8aToHex(test.addressRaw));

Public key: 0x0362cb81f6d8afe001df5b310e10b09f673c3399a63103ec6402ebbcba836619b7
SS58 public key 0x0d90aef20b4a64c8831c1420db3a2894d4e0d838d93acaacc3fec2f27b36db4b
```

When using `subkey` we can check the values
```sh
subkey inspect --scheme ecdsa "resource mirror lecture smooth midnight muffin position cup pepper fruit vanish also"
Secret phrase:       resource mirror lecture smooth midnight muffin position cup pepper fruit vanish also
  Network ID:        substrate
  Secret seed:       0x6e80514b03dcec9c289619fb49a2fdee158fc36b68d47ee3bf7a7c766960f142
  Public key (hex):  0x0362cb81f6d8afe001df5b310e10b09f673c3399a63103ec6402ebbcba836619b7
  Account ID:        0x0d90aef20b4a64c8831c1420db3a2894d4e0d838d93acaacc3fec2f27b36db4b
  Public key (SS58): KWAwr9VzcMSyi6AkY4CyC2bb9MziieDYeTtjnmmwXZAckBxeJ
  SS58 Address:      5CNVTcMkkxbBCBKviDvZG1JovUxj1EL2bSbu1NHBjf4djhEE
```

This is important because with other methods you could just pass their public key to a `MultiAddress` for example:
```dart
 final dest = $MultiAddress().id(hex.decode(publicKey));
```

But with ecdsa that would not be possible as the ecdsa public key is not used for that. In other crypto methods, sr25519 and ed25519 they always match that's why there is no problem with the other methods. You can see an example using sr25519 below:

```sh
subkey inspect "resource mirror lecture smooth midnight muffin position cup pepper fruit vanish also" 
Secret phrase:       resource mirror lecture smooth midnight muffin position cup pepper fruit vanish also
  Network ID:        substrate
  Secret seed:       0x6e80514b03dcec9c289619fb49a2fdee158fc36b68d47ee3bf7a7c766960f142
  Public key (hex):  0xbec6019e0194ba7c20dd706ca800dc61c749763bb4fb9464f4d655bce169d104
  Account ID:        0xbec6019e0194ba7c20dd706ca800dc61c749763bb4fb9464f4d655bce169d104
  Public key (SS58): 5GNqmvkkyXMZamAfTgRefneyAi7tZthypdkqCmRfuY3hc41D
  SS58 Address:      5GNqmvkkyXMZamAfTgRefneyAi7tZthypdkqCmRfuY3hc41D

```

**Comment by justkawal** [comment]
> ```js
> const keyring = new Keyring({ ss58Format: 42, type: 'ecdsa' });
> const test = keyring.addFromMnemonic('resource mirror lecture smooth midnight muffin position cup pepper fruit vanish also', {});
> console.log("Public key:", u8aToHex(test.publicKey));
> console.log('Account ID:', u8aToHex(test.addressRaw));
> ```

Done, Changed requested code.

**Comment by leonardocustodio** [comment]
Released under `secp256k1_ecdsa 0.1.0` and `polkadart_keyring 0.3.2`

**Pull request merged**